### PR TITLE
configfile: Add "include" support

### LINKED ISSUE #1145: Adding support for importing .cfg files into 'printer.cfg' ?
Examples speak louder than words, so lets assume you could include files and Klipper allowed you to redefine variables. A full printer.cfg could then look like this:

```
[include] 
 klipper/config/printer-creality-ender3-2018
 klipper/scripts/gcode_m300.cfg 
 custom_lcd_menu.cfg # complete rewrite of the menu config ~1200 lines long.

# reverse extruder and calibrate steps
[extruder]
 dir_pin: PB0
 step_distance: 0.011410184

[mcu]
serial: /dev/serial/by-id/usb-FTDI_FT232R_USB_UART_A107FTEV-if00-port0
```

But to be honest, I just want the ability to split my gcode scripts and lcd menu modifications to separate files as my printer.cfg is currently ~1200 line long and it's annoying me. Wouldn't need any redefines or anything, just allowing adding files to be appended to printer.cfg.

### PR DESCRIPTION
**Pull request by lauckhart**
Allows configuration files to include other configuration files using `[include filename.cfg]` syntax.  Klippy loads include files in the position of the include header; subsequent definitions override included values.

Supports wildcards (e.g. `[include macros/*.cfg]`).  Allows included files to include other files but blocks recursion.

Resolves #1145

Example `printer.cfg`:

```ini
# Load standard simulator config
[include klipper/config/avrsim.cfg]

# Make it faster
[printer]
max_velocity: 750
max_accel: 4500

# Add our macros
[include macros/*.cfg]
```

Signed-off-by: Greg Lauckhart <greg@lauckhart.com>

### CODE REVIEW

**Comment by KevinOConnor** [comment]
Thanks!  In general, it looks good to me.  I'd like to run some tests and also see if there is any other feedback.  So, I'll look to commit in a few days.

-Kevin

**Comment by lauckhart** [comment]
Cool.  I forgot to mention, this screws slightly with autosave because the logic that comments out manual configuration won't modify included files.  Not sure of the preferred way to handle but here are some options:

1. Recurse into included files and comment as the code does now for the main config,
2. Change saved configuration so it's just a normal config block and require folks to manually remove saved options if they want to revert to manual config, or
3. Leave as is and document the behavior

Personally I'd lean toward option 2 because it keeps the semantics simple, but I'm not familiar with all of the use cases.  Regardless I can tackle if we decide which direction to go.

**Comment by theopensourcerer** [comment]
Oooh - great I missed this. Thanks for doing this work @lauckhart.

One thought for the autosave... How about if there was a simple block header in the main file, e.g. [autosave] to indicate where the autosave should be. Then presumably one could decide to load your own include files above or below this?

**Comment by KevinOConnor** [comment]
I think that if SAVE_CONFIG finds a changed option not in the main config file, then SAVE_CONFIG should not modify anything and report an error.

-Kevin

**Comment by lauckhart** [comment]
@KevinOConnor pushed the logic you suggested.

@theopensourcerer I think somehow demarcating autosave data so you can add overrides lower in the file makes sense if autosave semantics change.  For this feature though Kevin's approach allows for merge without getting into tangential weeds.

**Comment by KevinOConnor** [comment]
I got a chance to test this.  It definitely seems useful.  However, I did notice two minor issues and one notable issue.

The main issue is that SAVE_CONFIG does not appear to be working as intended.  It seems that SAVE_CONFIG will now write out the complete merged config file (removing include directives and manually including the full config content into printer.cfg) - instead, it should only update the content of printer.cfg.

I'll cover the two minor issues separately.

Thanks,
-Kevin

File: klippy/configfile.py [hunk 1]
```diff
@@ -149,33 +149,75 @@ def _strip_duplicates(self, data, config):
...
+        if not include_filenames and not glob.has_magic(include_glob):
+            # Empty set is OK if wildcard but not for direct file reference
+            raise error("Include file '%s' does not exist", include_glob)
+        include_filenames.sort()
+        for include_filename in include_filenames:
+            include_data = self._read_config_file(include_filename)
+            self._parse_config(include_data, include_filename, fileconfig,
+                               visited)
+        return include_filenames
+    def _parse_config(self, data, filename, fileconfig, visited = set()):
```
Comment (KevinOConnor) [inline klippy/configfile.py]: Assigning `visited = set()` here effectively makes visited a global variable and that breaks the RESTART command.  A new set (or just a dictionary) should be passed on each _parse_config() call.

File: klippy/configfile.py [hunk 2]
```diff
@@ -210,6 +252,15 @@ def set(self, section, option, value):
...
     def remove_section(self, section):
         self.autosave.fileconfig.remove_section(section)
+    def _disallow_include_conflicts(self, regular_data, cfgname, gcode):
+        config = self._build_config_wrapper(regular_data, cfgname)
+        for section in self.autosave.fileconfig.sections():
+            for option in self.autosave.fileconfig.options(section):
+                if config.fileconfig.has_option(section, option):
+                    msg = "SAVE_CONFIG section '%s' option '%s' conflicts " \
+                          " with included value" % (section, option)
+                    logging.exception(msg)
```
Comment (KevinOConnor) [inline klippy/configfile.py]: I think logging.exception() should only be called from an exception handler.   The following `raise gcode.error()` will already log the error message, so it shouldn't be necessary to explicitly log it from here.

Comment (lauckhart) [reply klippy/configfile.py]: Fixed.  Good that it works this way, I was actually emulating the way exceptions were reported in cmd_SAVE_CONFIG.  You want me to remove the other extra logging while I'm in there?

**Comment by lauckhart** [comment]
Thanks for reviewing.  I can't actually reproduce the issue with SAVE_CONFIG expanding includes in the replacement file.  It only modifies the text of the file currently on disk.  LMK if I'm missing something, my testing of SAVE_CONFIG is admittedly a bit contrived.

Should I squash or do you want all of the commits?

File: klippy/configfile.py (hunk 1 above)
Comment (lauckhart) [reply klippy/configfile.py]: Fixed this too

**Comment by KevinOConnor** [comment]
Thanks!

It was my earlier SAVE_CONFIG tests that were goofed - not the code.  Sorry for the confusion.

-Kevin

**Comment by AxMod3DPrint** [comment]
Just wanted to say thanks for this @lauckhart 

Means I can standardise the menu and macros across all 5 printers that I have.

**Comment by theopensourcerer** [comment]
@lauckhart Yes. A big thanks for this. I updated to the latest master at the weekend and it seems to work perfectly. My printer.cfg is now clean and just contains the following:

[include config.d/printer.cfg]
[include config.d/tmc2208-steppers.cfg]
[include config.d/drv8825-bmg-extruder.cfg]
[include config.d/heater-bed.cfg]
[include config.d/probe.cfg]
[include config.d/bed-mesh.cfg]
[include config.d/gcode-macros.cfg]

Plus the results of SAVE_CONFIG below it.

**Comment by Adrian-at-CrimsonAuzre** [comment]
Sorry for the bit of necro, but this is why I love Klipper.

>Huh, wouldn't it be cool if I could do X
>*Googles "Klipper X"*
> Neat, support was added last week!

**Pull request merged**